src/core/shape_inference/include/proposal_shape_inference.hpp

```diff
@@ -30,78 +32,70 @@ void infer_prop_shape(const OpType* op,
+    if (image_shape_ps.rank().is_static()) {
+        NODE_VALIDATION_CHECK(
+            op,
+            image_shape_ps.size() == 1 && (image_shape_ps[0].compatible(3) || image_shape_ps[0].compatible(4)),
+            "Image_shape must be 1-D tensor and has got 3 or 4 elements (image_shape_shape[0]",
+            image_shape_ps,
+            ").");
     }
 
-    auto out_dim = DimType{};
-
-    if (class_probs_ps.rank().is_static() && bbox_deltas_ps.rank().is_static()) {
-        OPENVINO_ASSERT(DimType::merge(out_dim, class_probs_ps[0], bbox_deltas_ps[0]));
-    } else if (class_probs_ps.rank().is_static()) {
-        out_dim = class_probs_ps[0];
-    } else if (bbox_deltas_ps.rank().is_static()) {
-        out_dim = bbox_deltas_ps[0];
+    TShape proposed_boxes_shape;
+    proposed_boxes_shape.reserve(2);
+
+    if (class_probs_ps.rank().is_static()) {
+        proposed_boxes_shape.push_back(class_probs_ps[0]);
+
+        if (bbox_deltas_ps.rank().is_static()) {
+            // check anchor count and batch number consistency
+            NODE_VALIDATION_CHECK(op,
+                                  bbox_deltas_ps[1].compatible(class_probs_ps[1] * 2),
+                                  "Anchor number inconsistent between class_probs (",
+                                  class_probs_ps[1] * 2,
+                                  "), and bbox_deltas (",
+                                  bbox_deltas_ps[1],
+                                  ").");
+
+            NODE_VALIDATION_CHECK(op,
+                                  TDim::merge(proposed_boxes_shape[0], proposed_boxes_shape[0], bbox_deltas_ps[0]),
+                                  "Batch size inconsistent between class_probs (",
+                                  class_probs_ps[0],
+                                  ") and bbox deltas (",
+                                  bbox_deltas_ps[0],
+                                  ").");
+        }
     } else {
-        out_dim = Dimension::dynamic();
+        proposed_boxes_shape.emplace_back(ov::util::dim::inf_bound);
     }
 
-    auto& proposed_boxes_shape = output_shapes[0];
-    proposed_boxes_shape.resize(2);
-    proposed_boxes_shape[0] = out_dim * op->get_attrs().post_nms_topn;
-    proposed_boxes_shape[1] = 5;
-}
-template <class T>
-void shape_infer(const ov::op::v0::Proposal* op, const std::vector<T>& input_shapes, std::vector<T>& output_shapes) {
-    NODE_VALIDATION_CHECK(op, input_shapes.size() == 3 && output_shapes.size() == 1);
-    ov::op::v0::infer_prop_shape(op, input_shapes, output_shapes);
+    proposed_boxes_shape[0] *= op->get_attrs().post_nms_topn;
+    proposed_boxes_shape.emplace_back(5);
+    return proposed_boxes_shape;
 }
+}  // namespace proposal
 
+namespace v0 {
+template <class TShape>
+std::vector<TShape> shape_infer(const Proposal* op,
+                                const std::vector<TShape>& input_shapes,
+                                const ITensorAccessor& tensor_accessor = make_tensor_accessor()) {
```

Comment:
Is there a need to create `tensor_accessor` even if it's not used?

Reply:
It was not possible with adding some classes to shape inference.
I agree if parameter not used should not be defined at all and I updated it.

Just for information the `make_tensor_accessor` is cheap and I've done improvement  to tensor accessor that it will even simpler is just return reference to existing single object (created once per program life), so using this empty accessor as default parameter is similar to `int=0`;